File: src/main/java/edu/tju/ista/llm4test/execute/TestExecutor.java
package edu.tju.ista.llm4test.execute;

import edu.tju.ista.llm4test.adapter.AdapterRegistry;
import edu.tju.ista.llm4test.adapter.ProjectAdapter;
import edu.tju.ista.llm4test.adapter.jdk.JdkProjectAdapter;
import edu.tju.ista.llm4test.utils.LoggerUtil;
import edu.tju.ista.llm4test.concurrent.ConcurrentExecutionManager;

import java.io.File;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.logging.Level;


/**
 * 测试执行器（外观层）。
 * <p>
 * JDK/jtreg 的执行实现已迁移至 {@link JdkProjectAdapter}；
 * 本类保持原有公开API不变，委托适配器执行，并负责基于
 * ConcurrentExecutionManager 的异步提交（与harness无关的通用设施）。
 */
public class TestExecutor {
    private final ConcurrentExecutionManager concurrentManager;
    private final ProjectAdapter adapter;

    public TestExecutor() {
        this.concurrentManager = ConcurrentExecutionManager.getInstance();
        this.adapter = AdapterRegistry.get();
        // 环境检查
        adapter.checkEnvironment();
    }

    /**
     * 执行测试（主入口）
     * @param file 测试文件
     * @return 测试结果
     */
    public TestResult executeTest(File file) {
        return differentialTesting(file);
    }

    /**
     * 执行测试（使用TestCase）
     * @param testCase 测试用例
     * @return 测试结果
     */
    public TestResult executeTest(TestCase testCase) {
        return differentialTesting(testCase);
    }

    /**
     * 异步执行测试
     * @param file 测试文件
     * @return 测试结果的CompletableFuture
     */
    public CompletableFuture<TestResult> executeTestAsync(File file) {
        return concurrentManager.submitBatchTask(() -> differentialTesting(file));
    }

    /**
     * 差异化测试 - 在多个JDK上运行测试并比较结果（同步版本）
     * @param file 测试文件
     * @return 测试结果
     */
    public TestResult differentialTesting(File file) {
        TestCase testCase = new TestCase(file);
        return differentialTesting(testCase);
    }

    /**
     * 差异化测试 - 在多个JDK上运行测试并比较结果（使用TestCase）
     * @param testCase 测试用例
     * @return 测试结果
     */
    public TestResult differentialTesting(TestCase testCase) {
        return adapter.executeTest(testCase);
    }

    /**
     * 测试JDK环境（env 命令入口）
     */
    public void testJDKEnvironment() {
        adapter.verifyEnvironment();
    }

    /**
     * 清理临时目录（谨慎使用！）
     */
    public void clearTempDirectories() {
        // 由适配器自己决定清什么（接口上的默认实现是不做事），门面不再向下转型
        adapter.cleanupWorkspace();
    }

    /**
     * 关闭执行器并清理资源
     * 注意：这里不关闭并发管理器，因为它是单例且可能被其他组件使用
     */
    public void shutdown() {
        LoggerUtil.logExec(Level.INFO, "TestExecutor关闭完成");
    }

    /**
     * 获取并发执行状态
     */
    public void logConcurrentStatus() {
        concurrentManager.logStatus();
    }
}

File: src/main/java/edu/tju/ista/llm4test/execute/TestExecutor.kt
package edu.tju.ista.llm4test.execute

import edu.tju.ista.llm4test.utils.LoggerUtil
import java.io.File
import java.io.InputStream
import java.util.logging.Level
import kotlinx.coroutines.*

class TestExecutor(private val jarPath: String, private val resultDir: File) {



    companion object {
        // 目前未使用，可按需求后续扩展，比如在启动 jtreg 时加入不同的 exec 参数
        private val execFlags: Array<String> = arrayOf("-Xint", "-Xcomp", "-Xmixed")
        private val env: MutableMap<String, String> = hashMapOf("LANG" to "en_US.UTF-8")
        private val Path = System.getenv("PATH").replace("/usr/lib/jvm/java-17-openjdk-amd64/bin:", "")
        private val JDKs = arrayListOf("/home/Java/HotSpot/jdk-17.0.14+7", "/home/Java/HotSpot/jdk-21.0.6+7",
                "/home/Java/OpenJ9/jdk-17.0.14+7", "/home/Java/OpenJ9/jdk-21.0.6+7")
    }


    fun executeTest(file: File): TestResult {
        return try {
            val output = runBlocking { runJtreg(file) }
            var result = TestResult(output)
            for (i in 1..3) { // retry 3 times if fail
                if (result.isFail()) {
                    result = TestResult(output)
                } else {
                    break
                }
            }

            result
        } catch (e: Exception) {
            LoggerUtil.logExec(Level.WARNING, "Execute: ${file.path}\n${e.message}")
            e.printStackTrace()
            TestResult(TestResultKind.UNKNOWN)
        }
    }

    fun differentialTesting(file: File): TestResult {
        val results : HashMap<String, TestOutput> = hashMapOf()
        for (jdk in JDKs) {
            var jdk_env = HashMap(env)
            jdk_env["PATH"] = "$jdk/bin:$Path"
            val result = runBlocking { runJtreg(file, jdk_env) }
            results[jdk] = result
        }
        val res = TestResult()
        res.mergeResults(results)
        return res
    }

    @Throws(Exception::class)
    private fun readStream(stream: InputStream): String {
        return stream.bufferedReader().use { it.readText() }
    }

    @Throws(Exception::class)
    private suspend fun runJtreg(file: File) : TestOutput {
        return runJtreg(file, env)
    }

    @Throws(Exception::class)
    private suspend fun runJtreg(file: File,  env: MutableMap<String, String>): TestOutput = withContext(Dispatchers.IO) {
        // 清理临时文件
        clearJTworkFiles(file)

        // 构造 jtreg 命令及其参数，创建报告目录
        val reportDir = mkJReportDir()
        val jtregCommand = listOf("jtreg", "-avm", "-ea", "-va", "-r:$reportDir", file.path)
        LoggerUtil.logExec(Level.INFO, "execCommand: ${jtregCommand.joinToString(" ")}")

        // 使用 ProcessBuilder 构造进程并设置环境变量
        val processBuilder = ProcessBuilder(jtregCommand)
        processBuilder.environment().putAll(env)
        val process = processBuilder.start()

        // 使用 coroutine 并发消费标准输出和错误流
        val stdoutDeferred = async { readStream(process.inputStream) }
        val stderrDeferred = async { readStream(process.errorStream) }

        val finished = withTimeoutOrNull(600_000) { process.waitFor() }
        if (finished == null) {
            if (process.isAlive) {
                process.destroyForcibly()
            }
            LoggerUtil.logExec(Level.INFO, "Run: ${file.path}\n${jtregCommand.joinToString(" ")}\ntimeout")
            val stdout = stdoutDeferred.await()
            val stderr = stderrDeferred.await()
            return@withContext TestOutput(stdout, stderr, 124)
        }

        val exitValue = process.exitValue()
        val stdout = stdoutDeferred.await()
        val stderr = stderrDeferred.await()
        val output = TestOutput(stdout, stderr, exitValue)

        if (exitValue != 0) {
            LoggerUtil.logExec(
                    Level.SEVERE,
                    "TRY DEBUG THIS: -------------------------\n${file.path}\n${jtregCommand.joinToString(" ")}\n$output\n-------------------------------\n"
            )
        } else {
            LoggerUtil.logExec(Level.INFO, "Run: ${file.path}\n${jtregCommand.joinToString(" ")}\n$output")
        }
        output
    }

    fun clearJTworkFiles(file: File) {
        try {
            val baseName = file.name.removeSuffix(".java")
            val jtWork = File("JTwork")
            // 计算相对于 resultDir 的路径，需要确保 parent 不为 null
            val relativePath = resultDir.toPath().toAbsolutePath()
                    .relativize(file.toPath().toAbsolutePath()).parent ?: return

            val classCache = jtWork.toPath()
                    .resolve("classes")
                    .resolve(relativePath)
                    .resolve("$baseName.class")
                    .toFile()
            val jtrCache = jtWork.toPath()
                    .resolve(relativePath)
                    .resolve("$baseName.jtr")
                    .toFile()
            val dCache = jtWork.toPath()
                    .resolve(relativePath)
                    .resolve("$baseName.d")
                    .toFile()

            classCache.delete()
            jtrCache.delete()
            dCache.delete()
        } catch (e: Exception) {
            // 出现异常时忽略，可根据需要增加日志输出
        }
    }

    private fun mkJReportDir(): String {
        val threadId = Thread.currentThread().id
        val reportDir = File("tmp/tmp_${threadId}_${System.nanoTime()}")
        if (!reportDir.exists()) {
            reportDir.mkdirs()
        }
        return reportDir.path
    }

    public fun testJDKenv() {
        runBlocking {
            for (jdk in JDKs) {
                var jdk_env = HashMap(env)
                jdk_env["PATH"] = "$jdk/bin:$Path"
                val testCMD = listOf("which", "java")

                // 使用 ProcessBuilder 构造进程并设置环境变量
                val processBuilder = ProcessBuilder(testCMD)
                processBuilder.environment().putAll(jdk_env)
                val process = processBuilder.start()

                val stdoutDeferred = async { readStream(process.inputStream) }
                val stderrDeferred = async { readStream(process.errorStream) }

                val finished = withTimeoutOrNull(600_000) { process.waitFor() }
                if (finished == null) {
                    if (process.isAlive) {
                        process.destroyForcibly()
                    }
                    val stdout = stdoutDeferred.await()
                    val stderr = stderrDeferred.await()
                    System.out.println("TEST FAILED")
                }

                val exitValue = process.exitValue()
                val stdout = stdoutDeferred.await()
                val stderr = stderrDeferred.await()
                System.out.println("Execute Success")
                System.out.println(stdout)
                System.out.println(stderr)
                System.out.println("FINISH for " + jdk);
            }
        }

    }
}